src/admin/metrics.rs

```diff
@@ -0,0 +1,336 @@
+            // incoming.requests_complete(path, method, status, duration);
+        }
+    }
+}
+/// Accepts metrics request, encodes and prints
+#[cfg(feature = "prometheusd")]
+pub async fn accept_metrics_request(
+    tx: Request<hyper::body::Incoming>,
+    metrics_tx: Arc<RwLock<MetricSender>>,
+    registry_state: Arc<RwLock<StorageRegistry>>,
+) -> Result<hyper::Response<Full<Bytes>>, Infallible> {
+    // if tx.uri().path() == "/ws_metrics" {
+    //     return accept_ws_metrics(metrics_tx).await;
+    // } else if tx.uri().path() == "/http_metrics" {
+    //     return accept_http_metrics(metrics_tx).await;
+    // }
+    use crate::balancer::format::incoming_to_value;
+    use serde_json::{
+        json,
+        Value,
+        Value::Null,
+    };
+    // let mut tx = incoming_to_value(tx).await;
+    //     tx = json!({
+    //         "path": tx["path"],
+    //         "method": tx["method"],
+    //         "status": tx["status"],
+    //         "duration": tx["duration"],
+    //     });
+
+    let metrics_report = metrics_encoder(registry_state).await;
+    let response = Ok(hyper::Response::builder()
+        .status(200)
+        .body(Full::from(Bytes::from(metrics_report)))
+        .unwrap());
+    (response)
+}
+#[cfg(feature = "prometheusd")]
+async fn accept_ws_metrics(
+    _metrics_tx: Arc<MetricSender>,
+) -> Result<hyper::Response<Full<Bytes>>, Infallible> {
+    unimplemented!()
+}
+#[cfg(feature = "prometheusd")]
+async fn accept_http_metrics(
+    _metrics_tx: Arc<MetricSender>,
+) -> Result<hyper::Response<Full<Bytes>>, Infallible> {
+    unimplemented!()
+}
+#[cfg(feature = "prometheusd")]
+async fn metrics_encoder(storage_registry: Arc<RwLock<StorageRegistry>>) -> String {
+    use prometheus::Encoder;
+    let encoder = prometheus::TextEncoder::new();
+    let mut buffer = vec![];
+    let registry = storage_registry.read().unwrap().gather();
+    encoder.encode(&registry, &mut buffer).unwrap();
+    String::from_utf8(buffer).unwrap()
+}
+#[cfg(feature = "prometheusd")]
+pub async fn metrics_monitor(
+    metrics_rx: MetricReceiver,
+    storage_registry: Arc<RwLock<StorageRegistry>>,
+) {
+    //TODO: figure ownership mess here
+    let registry;
+    {
+        let registry_guard = storage_registry.read().unwrap();
+        registry = registry_guard;
+    }
+    let metrics_status = Arc::new(RwLock::new(
+        RpcMetrics::instance(&registry).unwrap().to_owned(),
+    ));
+    let metrics_stat_listener = metrics_status.clone();
+    tokio::spawn(metrics_listener(metrics_rx, metrics_stat_listener));
+    //metrics_processor(metrics_rx, metrics_status, "test", "test", &200, Duration::from_millis(100)).await;
+}
+// #[cfg(feature = "prometheusd")]
+pub async fn metrics_channel() -> (MetricSender, MetricReceiver) {
+    let (tx, rx) = unbounded_channel();
+    (tx, rx)
+}
+
+#[cfg(feature = "prometheusd")]
+pub async fn close(rx: &mut MetricReceiver) {
```

Comment:
why would we need to close the rx channel?

Reply:
I think I was hoping to have a flush and drop fns alongside close fn to help clear up metrics if needed for memory or whatever reasons. Not sure if it's really needed